Scripts/Runtime/Data/Collections/UnsafeTypedStream.cs

```diff
@@ -765,6 +825,77 @@ internal void ReadBlock(byte* dstStart, int dstStartIndex, int elementsToRead)
                 return ref UnsafeUtility.AsRef<T>(m_ReaderHead);
             }
         }
+        
+        //*************************************************************************************************************
+        // ENUMERATOR
+        //*************************************************************************************************************
+        
+        [BurstCompile]
+        public struct Enumerator : IEnumerator<T>
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            private int m_LaneIndex;
+            private LaneReader m_LaneReader;
+            private int m_ArrayIndex;
+            
+            public T Current
+            {
+                get => m_LaneReader.Read();
+            }
+            
+            internal Enumerator(BufferInfo* bufferInfo)
+            {
+                m_BufferInfo = bufferInfo;
+                m_LaneIndex = 0;
+                m_ArrayIndex = -1;
+                m_LaneReader = new LaneReader(m_BufferInfo, m_LaneIndex);
+            }
+
+            public bool MoveNext()
+            {
+                //Progress the array index
+                m_ArrayIndex++;
+                //So long as our lane reader has an entry we're good
+                if (m_ArrayIndex < m_LaneReader.Count)
+                {
+                    return true;
+                }
+                
+                //Otherwise we need to find the next lane that has something
+                m_LaneIndex++;
+                for (; m_LaneIndex < m_BufferInfo->LaneCount; ++m_LaneIndex)
+                {
+                    //Update lane reader and array index
+                    m_LaneReader = new LaneReader(m_BufferInfo, m_LaneIndex);
+                    m_ArrayIndex = 0;
+                    //If we find a lane that has something we're good
+                    if (m_ArrayIndex < m_LaneReader.Count)
+                    {
+                        return true;
+                    }
+                }
+                
+                //We've gone through all lanes and nothing is left, we're done
+                return false;
+            }
+
+            public void Reset()
+            {
+                m_LaneIndex = 0;
+                m_ArrayIndex = -1;
+                m_LaneReader = new LaneReader(m_BufferInfo, m_LaneIndex);
+            }
+
+            object IEnumerator.Current
+            {
+                get => throw new NotSupportedException($"Calling generic accessor which will box!");
+            }
```

Comment:
I wonder if you could burst discard this and only allow it when burst is disabled?
Not that it's a good idea but just for completion. Maybe for debug purposes you're dealing with these enumerators in a generic way 🤷

Reply:
Hmmmn, yeah I think that's a good idea. Changed.